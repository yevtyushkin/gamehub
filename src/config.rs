use figment::providers::Env;
use serde::*;

/// Application configuration.
#[derive(Debug, Deserialize, Clone, PartialEq)]
pub struct Config {
    /// Server configuration.
    pub server: ServerConfig,
    /// Postgres configuration.
    pub postgres: PostgresConfig,
}

impl Config {
    /// Loads [Config] from environment variables.
    #[allow(clippy::result_large_err)]
    pub fn from_env() -> figment::Result<Config> {
        figment::Figment::new()
            .merge(Env::raw().split("__"))
            .extract()
    }
}

/// Server configuration.
#[derive(Debug, Deserialize, Clone, PartialEq)]
pub struct ServerConfig {
    /// Server host address.
    pub host: String,
    /// Server port.
    pub port: u16,
}

impl ServerConfig {
    /// Returns server listening address as `host:port`.
    pub fn listen_addr(&self) -> String {
        format!("{}:{}", self.host, self.port)
    }
}

/// Postgres configuration.
#[derive(derive_more::Debug, Deserialize, Clone, PartialEq)]
pub struct PostgresConfig {
    /// Postgres host.
    pub host: String,
    /// Postgres port.
    pub port: u16,
    /// Postgres username.
    #[debug("<postgres_username_redacted>")]
    pub username: String,
    /// Postgres password.
    #[debug("<postgres_password_redacted>")]
    pub password: String,
    /// Postgres database name.
    pub database: String,
}

impl PostgresConfig {
    /// Returns Postgres connection URL as a string.
    pub fn connection_url(&self) -> String {
        format!(
            "postgres://{}:{}@{}:{}/{}",
            self.username, self.password, self.host, self.port, self.database
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn config_from_env() {
        figment::Jail::expect_with(|j| {
            j.set_env("SERVER__HOST", "127.0.0.1");
            j.set_env("SERVER__PORT", "8080");

            j.set_env("POSTGRES__HOST", "127.0.0.1");
            j.set_env("POSTGRES__PORT", "5432");
            j.set_env("POSTGRES__USERNAME", "postgres");
            j.set_env("POSTGRES__PASSWORD", "postgres_password");
            j.set_env("POSTGRES__DATABASE", "postgres_database");

            let config = Config::from_env()?;
            assert_eq!(
                config,
                Config {
                    server: ServerConfig {
                        host: "127.0.0.1".to_string(),
                        port: 8080
                    },
                    postgres: PostgresConfig {
                        host: "127.0.0.1".to_string(),
                        port: 5432,
                        username: "postgres".to_string(),
                        password: "postgres_password".to_string(),
                        database: "postgres_database".to_string(),
                    }
                }
            );

            Ok(())
        });
    }

    #[test]
    fn server_config_listen_addr() {
        let config = ServerConfig {
            host: "127.0.0.1".to_string(),
            port: 8080,
        };

        assert_eq!(config.listen_addr(), "127.0.0.1:8080");
    }

    #[test]
    fn postgres_config_connection_url() {
        let config = PostgresConfig {
            host: "127.0.0.1".to_string(),
            port: 5432,
            username: "postgres_username".to_string(),
            password: "postgres_password".to_string(),
            database: "postgres_database".to_string(),
        };

        assert_eq!(
            config.connection_url(),
            "postgres://postgres_username:postgres_password@127.0.0.1:5432/postgres_database"
        );
    }

    #[test]
    fn config_debug_does_not_leak_sensitive_info() {
        let config = Config {
            server: ServerConfig {
                host: "127.0.0.1".to_string(),
                port: 8080,
            },
            postgres: PostgresConfig {
                host: "127.0.0.1".to_string(),
                port: 5432,
                username: "postgres_1q2w3e4r_username".to_string(),
                password: "postgres_1q2w3e4r_password".to_string(),
                database: "postgres_database".to_string(),
            },
        };

        let debug = format!("{config:?}");

        assert!(!debug.contains("postgres_1q2w3e4r_username"));
        assert!(debug.contains("<postgres_username_redacted>"));

        assert!(!debug.contains("postgres_1q2w3e4r_password"));
        assert!(debug.contains("<postgres_password_redacted>"));
    }
}
