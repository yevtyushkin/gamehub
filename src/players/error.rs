use crate::api_error::ApiError;
use axum::Json;
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};

/// Possible players errors.
#[derive(Debug, thiserror::Error)]
pub enum PlayersError {
    /// Player not found.
    #[error("player not found")]
    PlayerNotFound,

    /// Error when verifying third party id token.
    #[error("error when verifying third party id token: {0}")]
    IdToken(#[from] id_token_verifier::IdTokenVerifierError),

    /// Error when verifying auth token token.
    #[error("error when verifying auth token: {0}")]
    AuthToken(jsonwebtoken::errors::Error),

    /// When auth token is missing.
    #[error("auth token is missing")]
    AuthTokenMissing,

    /// Internal error.
    #[error("internal error: {0}")]
    Internal(#[from] anyhow::Error),
}

impl IntoResponse for PlayersError {
    fn into_response(self) -> Response {
        let (status, id) = match &self {
            PlayersError::IdToken(_) => (StatusCode::BAD_REQUEST, 0),
            PlayersError::PlayerNotFound => (StatusCode::UNAUTHORIZED, 1),
            PlayersError::AuthToken(_) => (StatusCode::UNAUTHORIZED, 2),
            PlayersError::AuthTokenMissing => (StatusCode::UNAUTHORIZED, 3),
            PlayersError::Internal(_) => (StatusCode::INTERNAL_SERVER_ERROR, 4),
        };

        let body = ApiError {
            module: "players".into(),
            id,
            dev_message: self.to_string().into(),
        };

        (status, Json(body)).into_response()
    }
}
