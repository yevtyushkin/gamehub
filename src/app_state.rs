use crate::players::jwt_service::{JwtService, JwtServiceDefault};
use crate::players::players_service::{PlayersService, PlayersServiceDefault};
use std::sync::Arc;

/// Application state, shared between HTTP handlers.
pub trait AppState: Clone + Send + Sync + 'static {
    /// [PlayersService] implementation.
    type PlayersService: PlayersService + Send + Sync + 'static;
    /// Returns a reference to [Self::PlayersService] implementation.
    fn players_service(&self) -> &Self::PlayersService;

    /// [JwtService] implementation.
    type JwtService: JwtService + Send + Sync + 'static;
    /// Returns a reference to [Self::JwtService] implementation.
    fn jwt_service(&self) -> &Self::JwtService;
}

/// Default [AppState] implementation.
#[derive(Default)]
#[cfg_attr(test, derive(getset::WithSetters))]
#[cfg_attr(test, getset(set_with = "pub"))]
pub struct AppStateDefault<PS = PlayersServiceDefault, JS = JwtServiceDefault> {
    /// [PlayersService] implementation.
    players_service: PS,

    /// [JwtService] implementation.
    jwt_service: JS,
}

#[cfg(test)]
pub type MockAppState = AppStateDefault<
    crate::players::players_service::MockPlayersService,
    crate::players::jwt_service::MockJwtService,
>;

impl<PS, JS> AppStateDefault<PS, JS> {
    /// Creates a new [AppStateDefault] with the given services.
    pub fn new(players_service: PS, jwt_service: JS) -> AppStateDefault<PS, JS> {
        AppStateDefault {
            players_service,
            jwt_service,
        }
    }
}

impl<PS, JS> AppState for Arc<AppStateDefault<PS, JS>>
where
    PS: PlayersService + Send + Sync + 'static,
    JS: JwtService + Send + Sync + 'static,
{
    type PlayersService = PS;
    fn players_service(&self) -> &Self::PlayersService {
        &self.players_service
    }

    type JwtService = JS;
    fn jwt_service(&self) -> &Self::JwtService {
        &self.jwt_service
    }
}
