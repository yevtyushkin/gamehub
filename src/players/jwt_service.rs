use crate::app_state::AppState;
use crate::config::JwtConfig;
use crate::players::error::PlayersError;
use crate::players::player::PlayerId;
use anyhow::Context;
use axum::extract::FromRequestParts;
use axum::http::header::AUTHORIZATION;
use axum::http::request::Parts;
use chrono::{Duration, Utc};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;

/// Operations with [Player]s JWT tokens.
#[cfg_attr(test, mockall::automock)]
pub trait JwtService {
    /// Creates an [AuthToken] from the given [PlayerId].
    fn create_token(&self, player_id: PlayerId) -> Result<AuthToken<'static>, PlayersError>;

    /// Verifies the given [AuthToken] and returns its [AuthTokenClaims].
    #[allow(clippy::needless_lifetimes)]
    fn verify_token<'a>(&self, token: &AuthToken<'a>) -> Result<AuthTokenClaims, PlayersError>;
}

/// Auth token representing a result of a successful sign in.
#[derive(Debug, Clone, Serialize, Deserialize, derive_more::AsRef, PartialEq)]
pub struct AuthToken<'a>(pub Cow<'a, str>);

impl AuthToken<'_> {
    #[cfg(test)]
    /// Returns a test [AuthToken].
    pub fn test() -> AuthToken<'static> {
        AuthToken(Cow::Owned("auth_token".into()))
    }
}

/// [AuthToken] claims shape.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthTokenClaims {
    /// When the token expires.
    pub exp: i64,

    /// When the token was issued.
    pub iat: i64,

    /// Subject [PlayerId] the token is issued for.
    pub sub: PlayerId,
}

impl AuthTokenClaims {
    #[cfg(test)]
    /// Returns a test [AuthTokenClaims].
    pub fn test() -> AuthTokenClaims {
        AuthTokenClaims {
            exp: 123,
            iat: 456,
            sub: PlayerId::test(),
        }
    }
}

impl<S: AppState> FromRequestParts<S> for AuthTokenClaims {
    type Rejection = PlayersError;

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        match parts
            .headers
            .get(AUTHORIZATION)
            .and_then(|header_value| header_value.to_str().ok())
        {
            Some(header_value_str) => {
                let jwt_token = AuthToken(Cow::Borrowed(&header_value_str["Bearer ".len()..]));
                let claims = state.jwt_service().verify_token(&jwt_token)?;
                Ok(claims)
            }

            None => Err(PlayersError::AuthTokenMissing),
        }
    }
}

/// Default implementation of [JwtService].
#[derive(Clone)]
pub struct JwtServiceDefault {
    /// JWT [Validation] settings.
    validation: jsonwebtoken::Validation,

    /// JWT [DecodingKey] for signature verification.
    decoding_key: jsonwebtoken::DecodingKey,

    /// JWT [EncodingKey] for signing.
    encoding_key: jsonwebtoken::EncodingKey,

    /// JWT [Header] for JWT creation.
    header: jsonwebtoken::Header,

    /// [Duration] how long the token is valid for.
    token_ttl: Duration,
}

impl JwtServiceDefault {
    /// Creates a new [JwtServiceDefault] with the given [JwtConfig].
    pub fn new(config: JwtConfig) -> JwtServiceDefault {
        let validation = jsonwebtoken::Validation::default();
        let decoding_key = jsonwebtoken::DecodingKey::from_secret(config.secret.as_ref());
        let encoding_key = jsonwebtoken::EncodingKey::from_secret(config.secret.as_ref());
        let header = jsonwebtoken::Header::default();
        let token_ttl = config.ttl;

        JwtServiceDefault {
            validation,
            decoding_key,
            encoding_key,
            header,
            token_ttl,
        }
    }

    #[cfg(test)]
    /// Returns a test [JwtServiceDefault].
    pub fn test() -> JwtServiceDefault {
        JwtServiceDefault::new(JwtConfig::test())
    }
}

impl JwtService for JwtServiceDefault {
    fn create_token(&self, player_id: PlayerId) -> Result<AuthToken<'static>, PlayersError> {
        let now = Utc::now();

        let claims = AuthTokenClaims {
            exp: (now + self.token_ttl).timestamp(),
            iat: now.timestamp(),
            sub: player_id,
        };

        let token = jsonwebtoken::encode(&self.header, &claims, &self.encoding_key)
            .context("create auth token")?;

        Ok(AuthToken(Cow::Owned(token)))
    }

    fn verify_token(&self, token: &AuthToken) -> Result<AuthTokenClaims, PlayersError> {
        let claims = jsonwebtoken::decode(token.as_ref(), &self.decoding_key, &self.validation)
            .map_err(PlayersError::AuthToken)?
            .claims;

        Ok(claims)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn auth_token_claims_json_snapshot() {
        insta::assert_json_snapshot!(&AuthTokenClaims::test());
    }

    #[test]
    fn jwt_service_encode_and_decode_succeeds() -> anyhow::Result<()> {
        let player_id = PlayerId::test();
        let service = JwtServiceDefault::test();
        let ttl = service.token_ttl;

        let token = service.create_token(player_id)?;
        let claims = service.verify_token(&token)?;

        assert_eq!(claims.sub, player_id);
        assert!(claims.iat <= Utc::now().timestamp());
        assert_eq!(claims.iat + ttl.num_seconds(), claims.exp);

        Ok(())
    }

    #[test]
    fn jwt_service_decode_rejects_expired_tokens() -> anyhow::Result<()> {
        let service = JwtServiceDefault::test();
        let now = Utc::now();
        let exp_in_past = now.timestamp() - service.validation.leeway as i64 - 10;
        let iat_in_past = exp_in_past - service.token_ttl.num_seconds();
        let claims_expired = AuthToken(
            jsonwebtoken::encode(
                &service.header,
                &AuthTokenClaims {
                    exp: exp_in_past,
                    iat: iat_in_past,
                    sub: PlayerId::test(),
                },
                &service.encoding_key,
            )?
            .into(),
        );

        let result = service.verify_token(&claims_expired);

        assert!(
            matches!(result, Err(PlayersError::AuthToken(e)) if *e.kind() == jsonwebtoken::errors::ErrorKind::ExpiredSignature)
        );

        Ok(())
    }

    #[test]
    fn jwt_service_decode_rejects_wrong_signatures() -> anyhow::Result<()> {
        let service = JwtServiceDefault::test();
        let service_with_different_secret = {
            let mut service = service.clone();
            service.encoding_key =
                jsonwebtoken::EncodingKey::from_secret("wrong-secret".as_bytes());
            service
        };

        let claims_wrong_signature = service_with_different_secret.create_token(PlayerId::test())?;

        let result = service.verify_token(&claims_wrong_signature);
        assert!(
            matches!(result, Err(PlayersError::AuthToken(e)) if *e.kind() == jsonwebtoken::errors::ErrorKind::InvalidSignature)
        );

        Ok(())
    }
}
