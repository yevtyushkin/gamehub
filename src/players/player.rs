use crate::app_state::AppState;
use crate::players::error::PlayersError;
use crate::players::jwt_service::AuthTokenClaims;
use crate::players::players_service::PlayersService;
use axum::RequestPartsExt;
use axum::extract::FromRequestParts;
use axum::http::request::Parts;
use chrono::{DateTime, Utc};
use petname::{Generator, Petnames};
use serde::*;
use std::str::FromStr;
use uuid::Uuid;

/// Player representation in the application.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Player {
    /// [Player]'s ID.
    pub id: PlayerId,

    /// [Player]'s screen name.
    pub screen_name: PlayerScreenName,

    /// When a [Player] has joined.
    pub joined_at: PlayerJoinedAt,
}

impl Player {
    #[cfg(test)]
    /// Returns a test [Player].
    pub fn test() -> Player {
        Player {
            id: PlayerId::test(),
            screen_name: PlayerScreenName::test(),
            joined_at: PlayerJoinedAt::test(),
        }
    }
}

impl<S: AppState> FromRequestParts<S> for Player {
    type Rejection = PlayersError;
    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let claims: AuthTokenClaims = parts.extract_with_state(state).await?;
        let player = state.players_service().player_by_id(&claims.sub).await?;
        Ok(player)
    }
}

/// [Player]'s ID.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, sqlx::Type)]
pub struct PlayerId(pub Uuid);

impl PlayerId {
    /// Creates a new random [PlayerId].
    pub fn random() -> PlayerId {
        PlayerId(Uuid::now_v7())
    }

    #[cfg(test)]
    /// Returns a test [PlayerId].
    pub fn test() -> PlayerId {
        PlayerId(Uuid::from_u128(1234567890))
    }
}

/// [Player]'s screen name.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
pub struct PlayerScreenName(String);

impl PlayerScreenName {
    /// Creates a new random [PlayerScreenName].
    pub fn random() -> PlayerScreenName {
        let value = Petnames::small()
            .generate_one(3, "-")
            .unwrap_or_else(|| "no-screen-name".into());

        PlayerScreenName(value)
    }

    #[cfg(test)]
    /// Returns a test [PlayerScreenName].
    pub fn test() -> PlayerScreenName {
        PlayerScreenName("test-screen-name".into())
    }

    /// The maximum size of a [PlayerScreenName] in bytes.
    pub const MAX_SIZE: usize = 30;
}

/// [Player]'s screen name validation error.
#[derive(Debug, Clone, Copy, Eq, PartialEq, thiserror::Error)]
pub enum InvalidPlayerScreenName {
    /// [PlayerScreenName] is empty.
    #[error("player screen name is empty")]
    Empty,

    /// [PlayerScreenName] exceeds [PlayerScreenName::MAX_SIZE].
    #[error(
        "player screen name is larger than allowed max size of {} bytes",
        PlayerScreenName::MAX_SIZE
    )]
    ExceedsMaxSize,
}

impl FromStr for PlayerScreenName {
    type Err = InvalidPlayerScreenName;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.trim();

        if s.is_empty() {
            Err(InvalidPlayerScreenName::Empty)?
        }
        if s.len() > PlayerScreenName::MAX_SIZE {
            Err(InvalidPlayerScreenName::ExceedsMaxSize)?
        }

        Ok(PlayerScreenName(s.into()))
    }
}

/// When the [Player] joined.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, sqlx::Type, PartialEq)]
pub struct PlayerJoinedAt(pub DateTime<Utc>);

impl PlayerJoinedAt {
    /// Creates a new [PlayerJoinedAt] with the current time.
    pub fn now() -> PlayerJoinedAt {
        PlayerJoinedAt(Utc::now())
    }

    #[cfg(test)]
    /// Returns a test [PlayerJoinedAt].
    pub fn test() -> PlayerJoinedAt {
        PlayerJoinedAt(DateTime::UNIX_EPOCH)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn player_json_snapshot() {
        insta::assert_json_snapshot!(&Player::test());
    }

    #[test]
    fn player_screen_name_from_str_succeeds_if_screen_name_is_valid_of_min_size() {
        assert_eq!(
            PlayerScreenName::from_str("a"),
            Ok(PlayerScreenName("a".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str(" a  "),
            Ok(PlayerScreenName("a".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("г"),
            Ok(PlayerScreenName("г".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("  г     "),
            Ok(PlayerScreenName("г".into()))
        );
    }

    #[test]
    fn player_screen_name_from_str_succeeds_if_screen_name_is_valid_of_max_size() {
        let value = "w".repeat(PlayerScreenName::MAX_SIZE);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Ok(PlayerScreenName(value.clone()))
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Ok(PlayerScreenName(value))
        );

        let value = "г".repeat(PlayerScreenName::MAX_SIZE / 2);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Ok(PlayerScreenName(value.clone()))
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Ok(PlayerScreenName(value))
        );
    }

    #[test]
    fn player_screen_name_from_str_succeeds_if_screen_name_is_valid_of_medium_size() {
        assert_eq!(
            PlayerScreenName::from_str("abcd1234"),
            Ok(PlayerScreenName("abcd1234".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("  abcd1234     "),
            Ok(PlayerScreenName("abcd1234".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("абв123"),
            Ok(PlayerScreenName("абв123".into()))
        );

        assert_eq!(
            PlayerScreenName::from_str("  абв123     "),
            Ok(PlayerScreenName("абв123".into()))
        );
    }

    #[test]
    fn player_screen_name_from_str_fails_if_screen_name_is_empty() {
        assert_eq!(
            PlayerScreenName::from_str(""),
            Err(InvalidPlayerScreenName::Empty)
        );

        assert_eq!(
            PlayerScreenName::from_str("     "),
            Err(InvalidPlayerScreenName::Empty)
        );
    }

    #[test]
    fn player_screen_name_from_str_fails_if_screen_name_exceeds_max_size() {
        let value = "w".repeat(PlayerScreenName::MAX_SIZE + 1);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );

        let value = "г".repeat(PlayerScreenName::MAX_SIZE / 2 + 1);
        assert_eq!(
            PlayerScreenName::from_str(&value),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );

        let value_ws = format!("  {value}   ");
        assert_eq!(
            PlayerScreenName::from_str(&value_ws),
            Err(InvalidPlayerScreenName::ExceedsMaxSize)
        );
    }
}
