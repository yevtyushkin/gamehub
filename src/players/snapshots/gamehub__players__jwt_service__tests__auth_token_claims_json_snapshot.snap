---
source: src/players/jwt_service.rs
expression: "&AuthTokenClaims::test()"
---
{
  "exp": 123,
  "iat": 456,
  "sub": "00000000-0000-0000-0000-0000499602d2"
}
